baseplate/lib/prometheus_metrics.py

```diff
@@ -17,6 +18,146 @@
+)
+
+
+# http client labels and metrics
+http_client_latency_labels = [
+    "http_method",
+    "http_success",
+    "http_endpoint",
+    "http_slug",
+]
+http_client_requests_total_labels = [
+    "http_method",
+    "http_success",
+    "http_endpoint",
+    "http_response_code",
+    "http_slug",
+]
+http_client_active_requests_labels = [
+    "http_method",
+    "http_endpoint",
+    "http_slug",
+]
+
+http_client_latency_seconds = Histogram(
+    "http_client_latency_seconds",
+    "Description of histogram",
+    http_client_latency_labels,
+    buckets=default_buckets,
+)
+http_client_requests_total = Counter(
+    "http_client_requests_total",
+    "Description of counter",
+    http_client_requests_total_labels,
+)
+http_client_active_requests = Gauge(
+    "http_client_active_requests",
+    "Description of gauge",
+    http_client_active_requests_labels,
+)
+
+
+class PrometheusHTTPClientMetrics:
+    def __init__(self) -> None:
+        pass
+
+    def latency_seconds_metric(self, tags: Dict) -> Histogram:
+        return http_client_latency_seconds.labels(
+            http_method=tags.get("http.method", ""),
+            http_endpoint=getHTTPEndpointHandler(tags.get("http.url", "")),
+            http_success=tags.get("http.success", ""),
+            http_slug=tags.get("http.slug", ""),
+        )
+
+    def requests_total_metric(self, tags: Dict) -> Counter:
+        return http_client_requests_total.labels(
+            http_method=tags.get("http.method", ""),
+            http_endpoint=getHTTPEndpointHandler(tags.get("http.url", "")),
+            http_success=tags.get("http.success", ""),
+            http_response_code=tags.get("http.status_code", ""),
+            http_slug=tags.get("http.slug", ""),
+        )
+
+    def active_requests_metric(self, tags: Dict) -> Gauge:
+        return http_client_active_requests.labels(
+            http_method=tags.get("http.method", ""),
+            http_endpoint=getHTTPEndpointHandler(tags.get("http.url", "")),
+            http_slug=tags.get("http.slug", ""),
+        )
+
+    def get_latency_seconds_metric(self) -> Histogram:
+        return http_client_latency_seconds
+
+    def get_requests_total_metric(self) -> Counter:
+        return http_client_requests_total
+
+    def get_active_requests_metric(self) -> Gauge:
+        return http_client_active_requests
+
+
+def getHTTPEndpointHandler(httpURL: str) -> str:
+    """
+    hhtp://localhost:9090/user/123
+    Path to identify the endpoint handler, may be empty.
+    Ref: https://github.snooguts.net/reddit/baseplate.spec/blob/master/component-apis/prom-metrics.md#http
+    MUST NOT include the per-request value of "request fields":
+    For example, /subreddits/{subreddit_id}/posts/{post_id}, not /subreddits/1234/posts/5678.
+    """
+    from urllib3.util import parse_url
+
+    _, _, _, _, path, _, _ = parse_url(httpURL)
+    # TODO: remove request fields. Whats the best way to do this.
```

Comment:
Since these metrics wrap a generic HTTP client, they would have no semantic understanding of what's being called (i.e. what would correspond to a request field). Would be worth confirming with the author of the internal spec to ensure we're ok with this cardinality or if we'd rather opt for something with a lower cardinality like hostname, etc. (may already be captured by 'slug' - which is also an overloaded term 😞).

Reply:
that part is really more for the http server side of things
I just happened to copy/paste it because it worked, and in the end we should end up with something that does this on the server side.
But you're right, on the client side this doesn't apply and will just be a regular path.

Reply:
@MelissaCole might know more about how we could do this. I'll add her to the review.